src/functional/ffront/decorator.py

```diff
@@ -313,14 +313,24 @@ def format_itir(
         )
 
     def _validate_args(self, *args, **kwargs) -> None:
-        # TODO(tehrengruber): better error messages, check argument types
-        if len(args) != len(self.past_node.params):
-            raise GTTypeError(
-                f"Function takes {len(self.past_node.params)} arguments, but {len(args)} were given."
-            )
         if kwargs:
             raise NotImplementedError("Keyword arguments are not supported yet.")
 
+        arg_types = [symbol_makers.make_symbol_type_from_value(arg) for arg in args]
+        kwarg_types = {k: symbol_makers.make_symbol_type_from_value(v) for k, v in kwargs.items()}
+
+        try:
+            type_info.accepts_args(
+                self.past_node.type,
+                with_args=arg_types,
+                with_kwargs=kwarg_types,
+                raise_exception=True,
+            )
+        except GTTypeError as err:
+            raise ProgramTypeError.from_past_node(
```

Comment:
Maybe we should move `ProgramTypeError` somewhere else than the TypeDeduction. Not sure where though.